app/src/main/java/de/westnordost/streetcomplete/data/download/AVariableRadiusStrategy.kt

```diff
@@ -0,0 +1,72 @@
+import de.westnordost.osmapi.map.data.BoundingBox
+import de.westnordost.osmapi.map.data.LatLon
+import de.westnordost.streetcomplete.data.download.tiles.DownloadedTilesType
+import de.westnordost.streetcomplete.data.quest.VisibleQuestsSource
+import de.westnordost.streetcomplete.util.*
+import kotlin.math.PI
+import kotlin.math.max
+import kotlin.math.min
+import kotlin.math.sqrt
+
+/** Quest auto download strategy decides how big of an area to download based on the quest density */
+abstract class AVariableRadiusStrategy(
+    private val visibleQuestsSource: VisibleQuestsSource,
+    private val downloadedTilesDao: DownloadedTilesDao
+) : QuestAutoDownloadStrategy {
+
+    protected abstract val maxDownloadAreaInKm2: Double
+    protected abstract val desiredQuestCountInVicinity: Int
+
+    override fun getDownloadBoundingBox(pos: LatLon): BoundingBox? {
+        val tileZoom = ApplicationConstants.QUEST_TILE_ZOOM
+
+        val thisTile = pos.enclosingTile(tileZoom)
+        val hasMissingQuestsForThisTile = hasMissingQuestsFor(thisTile.toTilesRect())
+
+        // if at the location where we are, there is nothing yet, first download the tiniest
+        // possible bbox (~ 360x360m) so that we can estimate the quest density
+        if (hasMissingQuestsForThisTile) {
+            Log.i(TAG, "Downloading tiny area around user")
+            return thisTile.asBoundingBox(tileZoom)
+        }
+
+        // otherwise, see if anything is missing in a variable radius, based on quest density
+        val density = getQuestDensityFor(thisTile.asBoundingBox(tileZoom))
+        val radius = min(
+            sqrt( desiredQuestCountInVicinity / ( PI * density )),
+            sqrt( maxDownloadAreaInKm2 * 1000 * 1000 / PI )
+        )
+
+        val activeBoundingBox = pos.enclosingBoundingBox(radius)
+        if (hasMissingQuestsFor(activeBoundingBox.enclosingTilesRect(tileZoom))) {
+            Log.i(TAG, "Downloading in radius of ${radius.toInt()} meters around user")
+            return activeBoundingBox
+        }
+        Log.i(TAG, "All downloaded in radius of ${radius.toInt()} meters around user")
+        return null
+    }
+
+    /** return the quest density in quests per m² for this given [boundingBox]*/
+    private fun getQuestDensityFor(boundingBox: BoundingBox): Double {
+        val areaInKm = boundingBox.area()
+        val visibleQuestCount = visibleQuestsSource.getAllVisibleCount(boundingBox)
+        return visibleQuestCount / areaInKm
+    }
+
+    /** return if there are any quests in the given tiles rect that haven't been downloaded yet */
+    private fun hasMissingQuestsFor(tilesRect: TilesRect): Boolean {
+        val questExpirationTime = ApplicationConstants.REFRESH_QUESTS_AFTER
+        val ignoreOlderThan = max(0, System.currentTimeMillis() - questExpirationTime)
+        return !downloadedTilesDao.get(tilesRect, ignoreOlderThan).contains(DownloadedTilesType.QUESTS)
```

Comment:
`downloadedTilesDao` saves in which map tiles (at zoom 16) quests have already been downloaded. When trying to download quests in only tiles that have been recorded to have been downloaded already, download (manual or automatic) will not start, showing the message "No further quests available".

The lines you commented cause that tiles that have been registered to have been downloaded more than `REFRESH_QUESTS_AFTER` time ago (3 days) are not counted as downloaded anymore.

If every quest within the small area has been solved, the following things will happen:
1. if the `downloadedTilesDao` reports that this tile has not already been downloaded (less than 3 days ago), first the tiny area is downloaded again.
2. After that download finished (if any), the app will calculate that the quest density is 0 here. **This will lead to a division by zero error, gonna fix that now, thank you!** Since the quest quest density can then not be determined (is 0), the app will subsequently try to download the maximum allowed area by the download strategy (`maxDownloadAreaInKm2`). F.e. for the Wifi strategy it is 12 km². 
3. Before downloading however, it will check again if all the tiles that enclose the download area have already been downloaded (asking `downloadedTilesDao`). If they have, the message "No further quests available" will be shown.

Reply:
> Before downloading however, it will check again if all the tiles that enclose the download area have already been downloaded (asking `downloadedTilesDao`).

Got'cha. I didn't check the code outside this file so I wasn't sure if `downloadedTileDao` was keeping track of node download history or only the quests in them.

Reply:
Neither. It just records whether a download was done. :-)